### PR TITLE
[28] Implement post a jobs API @POST

### PR DESCRIPTION
**Pull request by badnameok**
## What did you do

- Implemented post jobs API
- Currently the job status will default to OPEN.
- Also includes a simple test unit.

## Demo

- Swagger -> Post /job

### CODE REVIEW

**Comment by badnameok** [comment]
มันบอกว่าไม่เจอ IsDateString แต่ใน branch เรามัน import ไว้แล้วหนิ :/

**Comment by PKhing** [comment]
where is all validation? ex. min age must <= max age , start time <= end time

**Comment by PKhing** [comment]
มี validation อื่นอีกครับ ไม่ได้มีแค่ที่พิมพ์บอกไปข้างบน

File: apps/api/src/modules/job/job.service.ts [hunk 1]
```diff
@@ -1,21 +1,43 @@
...
 export class JobService {
   constructor(private repository: JobRepository) {}
 
-  create(createJobDto: EditJobDto) {
-    return 'This action adds a new job'
+  async createJob(createJobDto: CreateJobDto, userId: number) {
+    // if minAge is greater than maxAge, throw error
+    if (createJobDto.minAge > createJobDto.maxAge) {
+      throw new ForbiddenException('minAge is greater than maxAge')
+      return
```
Comment (palmcm) [inline apps/api/src/modules/job/job.service.ts:25]:
```suggestion
      throw new BadRequestException('minAge is greater than maxAge')
```
If throw don't have to return and this is bad request error not forbidden

File: apps/api/src/modules/job/job.service.ts [hunk 2]
```diff
@@ -1,21 +1,43 @@
...
+      ) {
+        throw new ForbiddenException('startTime is greater than endTime')
+        return
+      }
+    }
+    try {
+      await this.repository.createJob(createJobDto, userId)
+    } catch (e) {
+      throw new InternalServerErrorException()
+    }
```
Comment (palmcm) [inline apps/api/src/modules/job/job.service.ts:40]: Should do something with e(error) like console.log(e). If not there is no need to try catch

File: packages/dtos/src/job.ts [hunk 3]
```diff
@@ -106,36 +108,121 @@ export class ShootingDto implements Partial<Shooting> {
...
   role: string
 
+  @IsNumber()
+  @IsOptional()
+  @IsNotEmpty()
   @ApiProperty()
   minAge: number
 
   @ApiProperty()
   maxAge: number
```
Comment (palmcm) [inline packages/dtos/src/job.ts:144]: Add decorator for this too

File: apps/api/src/modules/job/job.service.ts [hunk 4]
```diff
@@ -1,21 +1,43 @@
...
+    for (let i = 0; i < createJobDto.shooting.length; i++) {
+      if (
+        createJobDto.shooting[i].startTime > createJobDto.shooting[i].endTime
+      ) {
+        throw new ForbiddenException('startTime is greater than endTime')
+        return
+      }
+    }
+    try {
+      await this.repository.createJob(createJobDto, userId)
```
Comment (palmcm) [inline apps/api/src/modules/job/job.service.ts:37]: Should return created jobid

**Review by palmcm — CHANGES_REQUESTED** [review]
There will be more

File: apps/api/src/modules/job/job.controller.ts [hunk 5]
```diff
@@ -69,4 +60,15 @@ export class JobController {
   update(@Param('id') id: string, @Body() updateJobDto: EditJobDto) {
     return this.jobService.update(+id, updateJobDto)
   }
+
+  @Post()
+  @ApiCreatedResponse({ type: JobIdDTO })
+  @UseTypeAuthGuard('CASTING')
```
Comment (PKhing) [inline apps/api/src/modules/job/job.controller.ts:66]:
```suggestion
  @UseTypeAuthGuard(UserType.CASTING)
```

File: apps/api/src/modules/job/job.repository.ts [hunk 6]
```diff
@@ -1,12 +1,44 @@
...
+} from '@modela/dtos'
 import { Injectable } from '@nestjs/common'
 import { PrismaService } from 'src/database/prisma.service'
 
 @Injectable()
 export class JobRepository {
   constructor(private prisma: PrismaService) {}
 
+  async createJob(createJobDto: CreateJobDto, userId: number) {
+    const job = await this.prisma.job.create({
```
Comment (PKhing) [inline apps/api/src/modules/job/job.repository.ts:16]: image

File: apps/api/src/modules/job/job.repository.ts [hunk 7]
```diff
@@ -1,12 +1,44 @@
...
+        Shooting: {
+          create: createJobDto.shooting.map((shooting) => ({
+            startDate: shooting.startDate,
+            endDate: shooting.endDate,
+            startTime: shooting.startTime,
+            endTime: shooting.endTime,
+            shootingLocation: shooting.shootingLocation,
+          })),
+        },
+      },
```
Comment (PKhing) [inline apps/api/src/modules/job/job.repository.ts:38]:
```suggestion
     data: {
        status: JobStatus.OPEN,
        castingId: userId,
        ...rest,
        Shooting: {
          create: shooting,
        },
      },
```

File: apps/api/src/modules/job/job.service.spec.ts [hunk 8]
```diff
@@ -267,4 +268,37 @@ describe('JobService', () => {
...
+  describe('postJob', () => {
+    const MOCK_TITLE = 'test title'
+    const MOCK_DESCRIPTION = 'test description'
+    const MOCK_STATUS = JobStatus.OPEN
+    const MOCK_ROLE = 'test role'
+    const MOCK_AGE = 20
+    const MOCK_ACTORCOUNT = 1
+    const MOCK_WAGE = 10000
+    const MOCK_DEADLINE = new Date()
+    const MOCK_SHOOTING = mock('shooting').get(3)
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:281]: image

File: apps/api/src/modules/job/job.service.spec.ts [hunk 9]
```diff
@@ -267,4 +268,37 @@ describe('JobService', () => {
...
+        .omit(['castingId', 'createdAt', 'updatedAt', 'jobId', 'status'])
+        .get(),
+      castingId: MOCK_CASTING_ID,
+      shooting: mock('shooting').get(3),
+    }
+
+    describe('normal behavior', () => {
+      it('should create the job successfully', async () => {
+        jest.spyOn(repository, 'createJob').mockResolvedValue()
+        jest.spyOn(service, 'createJob').mockResolvedValue()
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:295]: Why did you mock return service? If you mock return that how can you test whether the return value is correct or not?

File: apps/api/src/modules/job/job.service.spec.ts [hunk 10]
```diff
@@ -267,4 +268,37 @@ describe('JobService', () => {
...
+
+        await service.createJob(MOCK_JOB, MOCK_CASTING_ID)
+
+        expect(repository.createJob).toBeCalledWith(MOCK_JOB, MOCK_CASTING_ID)
+        expect(service.createJob).toBeCalledWith(MOCK_JOB, MOCK_CASTING_ID)
```
Comment (PKhing) [inline apps/api/src/modules/job/job.service.spec.ts:300]: You called `service.createJob` 3 lines above. What do you want to test here?

**Review by PKhing — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/job/job.service.spec.ts [hunk 11]
```diff
@@ -1,11 +1,12 @@
-import { mock, UserType } from '@modela/database'
+import { JobStatus, mock, UserType } from '@modela/database'
 import {
   GetJobCardDto,
   GetJobCardWithMaxPageDto,
   SearchJobDto,
 } from '@modela/dtos'
 import { ForbiddenException, NotFoundException } from '@nestjs/common'
 import { Test, TestingModule } from '@nestjs/testing'
+import exp from 'constants'
```
Comment (miello) [inline apps/api/src/modules/job/job.service.spec.ts:9]: Consider remove this, unused import

**Review by miello — CHANGES_REQUESTED** [review]

File: packages/dtos/src/job.ts [hunk 12]
```diff
@@ -106,36 +108,121 @@ export class ShootingDto implements Partial<Shooting> {
 type EditJobType = Partial<Job> & { shooting: ShootingDto[] }
 
 export class EditJobDto implements EditJobType {
+
+  @IsString()
+  @IsOptional()
+  @IsNotEmpty()
```
Comment (miello) [inline packages/dtos/src/job.ts:114]: Is title either required or optional ?

Comment (miello) [reply packages/dtos/src/job.ts:114]: Please consider for all properties also

File: packages/dtos/src/job.ts [hunk 13]
```diff
@@ -106,36 +108,121 @@ export class ShootingDto implements Partial<Shooting> {
...
+  @ApiProperty()
+  wage: number
+
+
+  @IsDateString()
+  @IsNotEmpty()
   @ApiProperty()
   applicationDeadline: Date
 
+  @IsNotEmpty()
```
Comment (miello) [inline packages/dtos/src/job.ts:225]: Does we need to validate `ShootingDto` verify also ?

**Review by miello — CHANGES_REQUESTED** [review]

**Review by PKhing — CHANGES_REQUESTED** [review]
Consider these input
- minAge: -1
- actorCount: 3.5
- wage: -4235
- applicationDeadline: yesterday

File: apps/api/src/modules/job/job.service.ts [hunk 14]
```diff
@@ -9,16 +10,69 @@ import {
...
+  async createJob(createJobDto: CreateJobDto, userId: number) {
+    // if minAge is greater than maxAge, throw error
+    if (createJobDto.minAge > createJobDto.maxAge) {
+      throw new BadRequestException('minAge is greater than maxAge')
+    }
+    // for each shooting, if the start time is greater than the end time, throw error
+    for (let i = 0; i < createJobDto.shooting.length; i++) {
+      if (
+        createJobDto.shooting[i].startTime > createJobDto.shooting[i].endTime &&
+        createJobDto.shooting[i].startDate == createJobDto.shooting[i].endDate
```
Comment (miello) [inline apps/api/src/modules/job/job.service.ts:34]:
```suggestion
        createJobDto.shooting[i].startDate === createJobDto.shooting[i].endDate
```

File: apps/api/src/modules/job/job.service.spec.ts [hunk 15]
```diff
@@ -344,4 +349,156 @@ describe('JobService', () => {
...
+        }
+        if (MOCK_JOB.minAge < 1) {
+          MOCK_JOB.minAge = 1
+        }
+        if (MOCK_JOB.wage < 1) {
+          MOCK_JOB.wage = 1
+        }
+
+        const result = mock('job').get(1)[0]
+        console.log(result)
```
Comment (miello) [inline apps/api/src/modules/job/job.service.spec.ts:444]: Consider remove `console.log`

**Review by miello — APPROVED** [review]
I will ignore these for now please fix it in `jj/back/update-job-api`

**Pull request merged**